packages/lit-html/src/lib/lit-html.ts

```diff
@@ -649,10 +701,19 @@ export class AttributePart {
    * @param value the raw input value to normalize
    * @param _i the index in the values array this value was read from
    */
-  __resolveValue(value: unknown, _i: number) {
-    // TODO (justinfagnani): invoke directives here, which will need
-    // _i to revive directive state
-    return (value ??= '');
+  __resolveValue(value: unknown, i: number) {
+    const directiveCtor = (value as DirectiveResult)?._$litDirective$;
+    if (directiveCtor !== undefined) {
+      let directive: Directive = (this.__directives ??= [])[i];
```

Comment:
This essentially guarantees that the `__directives` array will go into slow mode because of the OOB access the first time. Can we initialize to a correct length to begin with?

Reply:
We can initialize to `new Array(this.__strings.length - 1)`. It'd be good to skip the array when `this.__strings === undefined`

Reply:
One thing that might complicate this soon is that in order to support nested directives we might have a longer array than the values array, and won't know the length until we see all the directives. I'll leave a TODO for now.